File: academic/cpp/src/lista_estoque.cpp
#include "lista_estoque.hpp"

ListaEstoque::ListaEstoque() : cabeca(nullptr), cauda(nullptr), quantidade(0) {}

ListaEstoque::~ListaEstoque() {
    limpar();
}

void ListaEstoque::inserir(const BolsaAed& bolsa) {
    NoBolsa* novo = new NoBolsa(bolsa);
    if (cabeca == nullptr) {
        cabeca = novo;
        cauda = novo;
    } else {
        cauda->proximo = novo;
        cauda = novo;
    }
    quantidade++;
}

bool ListaEstoque::buscarPorCodigo(const std::string& codigo, BolsaAed& encontrada) const {
    NoBolsa* atual = cabeca;
    while (atual != nullptr) {
        if (atual->dado.codigo == codigo) {
            encontrada = atual->dado;
            return true;
        }
        atual = atual->proximo;
    }
    return false;
}

bool ListaEstoque::removerPorCodigo(const std::string& codigo, BolsaAed& removida) {
    NoBolsa* anterior = nullptr;
    NoBolsa* atual = cabeca;
    while (atual != nullptr) {
        if (atual->dado.codigo == codigo) {
            if (anterior == nullptr) {
                cabeca = atual->proximo;
            } else {
                anterior->proximo = atual->proximo;
            }
            if (atual == cauda) {
                cauda = anterior;
            }
            removida = atual->dado;
            delete atual;
            quantidade--;
            return true;
        }
        anterior = atual;
        atual = atual->proximo;
    }
    return false;
}

bool ListaEstoque::primeiraCompativelPorIgualdade(const std::string& tipoSanguineo,
                                                  const std::string& componente,
                                                  const std::string& statusExigido,
                                                  BolsaAed& escolhida) const {
    NoBolsa* atual = cabeca;
    while (atual != nullptr) {
        const bool mesmoTipo = atual->dado.tipoSanguineo == tipoSanguineo;
        const bool mesmoComponente = atual->dado.componente == componente;
        const bool statusOk = statusExigido.empty() || atual->dado.status == statusExigido;
        if (mesmoTipo && mesmoComponente && statusOk) {
            escolhida = atual->dado;
            return true;
        }
        atual = atual->proximo;
    }
    return false;
}

int ListaEstoque::tamanho() const {
    return quantidade;
}

bool ListaEstoque::estaVazia() const {
    return cabeca == nullptr;
}

void ListaEstoque::limpar() {
    NoBolsa* atual = cabeca;
    while (atual != nullptr) {
        NoBolsa* proximo = atual->proximo;
        delete atual;
        atual = proximo;
    }
    cabeca = nullptr;
    cauda = nullptr;
    quantidade = 0;
}

File: src/main/java/br/edu/cesar/bloodell/aed/u1/ListaEstoque.java
package br.edu.cesar.bloodell.aed.u1;

import br.edu.cesar.bloodell.aed.contratos.BolsaAed;

public class ListaEstoque {

    private NoBolsa cabeca;
    private NoBolsa cauda;
    private int quantidade;

    public ListaEstoque() {
        this.cabeca = null;
        this.cauda = null;
        this.quantidade = 0;
    }

    public void inserir(BolsaAed bolsa) {
        if (bolsa == null) {
            throw new IllegalArgumentException("Nao e possivel inserir bolsa nula na lista.");
        }
        NoBolsa novo = new NoBolsa(bolsa);
        if (cabeca == null) {
            cabeca = novo;
            cauda = novo;
        } else {
            cauda.setProximo(novo);
            cauda = novo;
        }
        quantidade++;
    }

    public BolsaAed buscarPorCodigo(String codigo) {
        NoBolsa atual = cabeca;
        while (atual != null) {
            if (atual.getDado().getCodigo().equals(codigo)) {
                return atual.getDado();
            }
            atual = atual.getProximo();
        }
        return null;
    }

    public BolsaAed removerPorCodigo(String codigo) {
        NoBolsa anterior = null;
        NoBolsa atual = cabeca;
        while (atual != null) {
            if (atual.getDado().getCodigo().equals(codigo)) {
                if (anterior == null) {
                    cabeca = atual.getProximo();
                } else {
                    anterior.setProximo(atual.getProximo());
                }
                if (atual == cauda) {
                    cauda = anterior;
                }
                atual.setProximo(null);
                quantidade--;
                return atual.getDado();
            }
            anterior = atual;
            atual = atual.getProximo();
        }
        return null;
    }

    public BolsaAed primeiraCompativelPorIgualdade(String tipoSanguineo, String componente, String statusExigido) {
        NoBolsa atual = cabeca;
        while (atual != null) {
            BolsaAed bolsa = atual.getDado();
            boolean mesmoTipo = bolsa.getTipoSanguineo().equals(tipoSanguineo);
            boolean mesmoComponente = bolsa.getComponente().equals(componente);
            boolean statusOk = statusExigido == null || statusExigido.equals(bolsa.getStatus());
            if (mesmoTipo && mesmoComponente && statusOk) {
                return bolsa;
            }
            atual = atual.getProximo();
        }
        return null;
    }

    public BolsaAed[] paraVetor() {
        BolsaAed[] itens = new BolsaAed[quantidade];
        NoBolsa atual = cabeca;
        int indice = 0;
        while (atual != null) {
            itens[indice] = atual.getDado();
            indice++;
            atual = atual.getProximo();
        }
        return itens;
    }

    public int tamanho() {
        return quantidade;
    }

    public boolean estaVazia() {
        return cabeca == null;
    }

    public void limpar() {
        cabeca = null;
        cauda = null;
        quantidade = 0;
    }
}
